objects/src/builder/mod.rs

```diff
@@ -252,22 +254,34 @@ impl<T: Rng> AccountBuilder<T> {
         self
     }
 
-    pub fn code_root(&mut self, code_root: Digest) -> &mut Self {
-        self.account_id_builder.code_root(code_root);
-        self
-    }
-
-    pub fn storage_root(&mut self, storage_root: Digest) -> &mut Self {
-        self.account_id_builder.storage_root(storage_root);
-        self
-    }
-
     pub fn build(mut self) -> Result<Account, AccountError> {
-        let account_id = self.account_id_builder.build()?;
         let vault = AccountVault::new(&self.assets)?;
         let storage = self.storage_builder.build();
         let assembler = assembler();
+
         let account_module_ast = ModuleAst::parse(&self.code).unwrap();
+        let library_path = LibraryPath::new("builder::account").expect("valid path");
+        let module = Module::new(library_path, account_module_ast.clone());
+
+        let mut procedure_digests = assembler
+            .compile_module(&module, &mut AssemblyContext::new(AssemblyContextType::Module))
+            .map_err(AccountError::AccountCodeAssemblerError)?;
+        procedure_digests.sort_by_key(|a| a.as_bytes());
+
+        let procedure_tree = SimpleSmt::with_leaves(
+            ACCOUNT_CODE_TREE_DEPTH,
+            procedure_digests
+                .iter()
+                .enumerate()
+                .map(|(idx, p)| (idx as u64, p.into()))
+                .collect::<Vec<_>>(),
+        )
+        .unwrap();
+
+        self.account_id_builder.code_root(procedure_tree.root());
+        self.account_id_builder.storage_root(storage.root());
+        let account_id = self.account_id_builder.build()?;
+
         let account_code = AccountCode::new(account_id, account_module_ast, &assembler)?;
```

Comment:
much cleaner! changed

Reply:
maybe we should change `AccountId::new` interface to receive `Option<AccountId>` instead?

Reply:
I'm going to assume that you mean `AccountCode::new(..)` should take an `Option`.  Yeah I definitely think we could consider this. Some thoughts on this... given that the `Assembler` is provided from the client's context I think it should also be the responsibility of the client to determine the namespace against which the code is compiled.  Following this train of thought maybe instead of the client providing an `AccountId` maybe they should provide a `LibraryPath`.